File: android/app/src/main/java/com/vergipro/mobile/core/data/RepositoryEnvironment.kt
package com.vergipro.mobile.core.data

import com.vergipro.mobile.BuildConfig
import com.vergipro.mobile.core.network.OrganizationContext

enum class RepositoryEnvironment { Production, Preview }

fun RepositoryEnvironment.validated(): RepositoryEnvironment =
    if (BuildConfig.DEBUG) this else RepositoryEnvironment.Production

fun interface TenantScopedRepository<T> {
    suspend fun load(organization: OrganizationContext): T
}

class RepositoryContainer private constructor(
    val environment: RepositoryEnvironment,
) {
    companion object {
        fun create(environment: RepositoryEnvironment) =
            RepositoryContainer(environment.validated())
    }
}

File: ios/VergiPro/Core/Data/RepositoryEnvironment.swift
import Foundation

enum RepositoryEnvironment: Sendable {
    case production
    case preview

    static func validated(_ requested: RepositoryEnvironment) -> RepositoryEnvironment {
        #if DEBUG
        requested
        #else
        .production
        #endif
    }
}

protocol TenantScopedRepository: Sendable {
    associatedtype Value: Sendable

    func load(for organization: OrganizationContext) async throws -> Value
}

struct RepositoryContainer: Sendable {
    let environment: RepositoryEnvironment

    init(environment: RepositoryEnvironment) {
        self.environment = .validated(environment)
    }
}
